watchdog/src/health.rs

```diff
@@ -0,0 +1,204 @@
+use crate::bitcoin_block_apis::BitcoinBlockApi;
+use crate::fetch::BlockInfo;
+use candid::CandidType;
+use serde::{Deserialize, Serialize};
+
+/// Status codes for the health of Bitcoin canister.
+#[derive(Clone, Debug, CandidType, PartialEq, Eq, Serialize, Deserialize)]
+pub enum StatusCode {
+    #[serde(rename = "no_data")]
+    NoData,
+    #[serde(rename = "ok")]
+    Ok,
+    #[serde(rename = "ahead")]
+    Ahead,
+    #[serde(rename = "behind")]
+    Behind,
+}
+
+/// The health status of the Bitcoin canister.
+#[derive(Clone, Debug, CandidType, PartialEq, Eq, Serialize, Deserialize)]
+pub struct HealthStatus {
+    pub source_height: Option<u64>,
+    pub other_number: u64,
+    pub other_heights: Vec<u64>,
+    pub target_height: Option<u64>,
+    pub height_diff: Option<i64>,
+    pub status: StatusCode,
+}
+
+/// Calculates the health status of the Bitcoin canister.
+pub fn calculate() -> HealthStatus {
+    compare(
+        crate::storage::get(&BitcoinBlockApi::BitcoinCanister),
+        BitcoinBlockApi::explorers()
+            .iter()
+            .filter_map(crate::storage::get)
+            .collect::<Vec<_>>(),
+    )
+}
+
+/// Compares the source with the other explorers.
+fn compare(source: Option<BlockInfo>, other: Vec<BlockInfo>) -> HealthStatus {
+    let source_height = source.and_then(|block| block.height);
+    let heights = other
+        .iter()
+        .filter_map(|block| block.height)
+        .collect::<Vec<_>>();
+    let other_number = heights.len() as u64;
+    let other_heights = heights.clone();
+    let target_height = if heights.len() < crate::config::MIN_EXPLORERS {
+        None // Not enough data from explorers.
+    } else {
+        median(heights)
+    };
+    let height_diff = source_height
+        .zip(target_height)
+        .map(|(source, target)| source as i64 - target as i64);
+    let status = height_diff.map_or(StatusCode::NoData, |diff| {
+        if diff < crate::config::BLOCKS_BEHIND_THRESHOLD {
+            StatusCode::Behind
+        } else if diff > crate::config::BLOCKS_AHEAD_THRESHOLD {
+            StatusCode::Ahead
+        } else {
+            StatusCode::Ok
+        }
+    });
+
+    HealthStatus {
+        source_height,
+        other_number,
+        other_heights,
+        target_height,
+        height_diff,
+        status,
+    }
+}
+
+/// The median of the given values.
+fn median<T: std::cmp::Ord + Clone>(mut values: Vec<T>) -> Option<T> {
+    let n = values.len();
+    if n == 0 {
+        None
+    } else {
+        values.sort();
+        let mid = if n % 2 == 0 { (n - 1) / 2 } else { n / 2 };
```

Comment:
Technically, the definition of median for an even list of numbers is the average of the two middle values. Is there a reason not using this definition? I don't believe it matters much but just noticed it.

Reply:
Done.